File: apps/web_pwa/src/core/voice/voiceService.ts
import type { SpeechRecognitionLike } from './speechRecognition';

// Mirrors apps/mobile_flutter/lib/core/voice/voice_service.dart, ported
// from flutter_tts/speech_to_text to the browser's native
// speechSynthesis / SpeechRecognition APIs. Same stopgap status as the
// Flutter version: no guaranteed consistent voice per learner, and STT
// browser support is inconsistent (reliably Chrome/Chromium only) -- not a
// regression versus Flutter, just a different flavor of the same caveat.
// Stage 2/3 in docs/product/roadmap-learning-features.md replaces this with
// provider-gateway STT/TTS.

export type SpeakOutcome = 'spokenInLanguage' | 'spokenWithFallbackVoice' | 'noVoice';
export type VoiceListenOutcome = 'listening' | 'unavailable' | 'permissionDenied';

export class VoiceService {
  private recognition: SpeechRecognitionLike | null = null;
  private listening = false;

  async speak(text: string, localeId: string): Promise<SpeakOutcome> {
    if (!text.trim()) {
      return 'noVoice';
    }
    if (!('speechSynthesis' in window)) {
      return 'noVoice';
    }

    window.speechSynthesis.cancel();

    const hasMatchingVoice = await this.hasVoiceForLocale(localeId);

    return new Promise((resolve) => {
      const utterance = new SpeechSynthesisUtterance(text);
      utterance.lang = localeId;
      utterance.rate = 0.85;

      utterance.onend = () => resolve(hasMatchingVoice ? 'spokenInLanguage' : 'spokenWithFallbackVoice');
      utterance.onerror = () => resolve('noVoice');

      try {
        window.speechSynthesis.speak(utterance);
      } catch {
        resolve('noVoice');
      }
    });
  }

  stopSpeaking(): void {
    if ('speechSynthesis' in window) {
      window.speechSynthesis.cancel();
    }
  }

  /** Best-effort check for whether any installed voice matches `localeId`
   * (exact match) or at least its base language (e.g. "kn" for "kn-IN").
   * `getVoices()` can return an empty list on first call before the
   * browser's voice list has loaded, so this fails open (assumes no
   * dedicated voice) rather than blocking playback. */
  private async hasVoiceForLocale(localeId: string): Promise<boolean> {
    if (!('speechSynthesis' in window)) {
      return false;
    }
    try {
      const voices = window.speechSynthesis.getVoices();
      const languagePrefix = localeId.split('-')[0]?.toLowerCase() ?? '';
      return voices.some((voice) => {
        const voiceLocale = voice.lang?.toLowerCase();
        return voiceLocale === localeId.toLowerCase() || voiceLocale?.startsWith(`${languagePrefix}-`);
      });
    } catch {
      return false;
    }
  }

  async startListening(
    localeId: string,
    onResult: (recognizedWords: string, isFinal: boolean) => void,
  ): Promise<VoiceListenOutcome> {
    const SpeechRecognitionCtor = window.SpeechRecognition ?? window.webkitSpeechRecognition;
    if (!SpeechRecognitionCtor) {
      return 'unavailable';
    }

    try {
      const recognition = new SpeechRecognitionCtor();
      recognition.lang = localeId;
      recognition.continuous = false;
      recognition.interimResults = true;

      recognition.onresult = (event) => {
        let transcript = '';
        let isFinal = false;
        for (let i = event.resultIndex; i < event.results.length; i += 1) {
          const result = event.results[i];
          transcript += result[0].transcript;
          isFinal = isFinal || result.isFinal;
        }
        onResult(transcript, isFinal);
      };

      recognition.onerror = (event) => {
        this.listening = false;
        if (event.error === 'not-allowed' || event.error === 'permission-denied') {
          // Surfaced to the caller via the returned outcome for the initial
          // start() failure path; mid-session denials just stop listening.
        }
      };

      recognition.onend = () => {
        this.listening = false;
      };

      recognition.start();
      this.recognition = recognition;
      this.listening = true;
      return 'listening';
    } catch (error) {
      if (error instanceof DOMException && error.name === 'NotAllowedError') {
        return 'permissionDenied';
      }
      return 'unavailable';
    }
  }

  async stopListening(): Promise<void> {
    this.recognition?.stop();
    this.listening = false;
  }

  get isListening(): boolean {
    return this.listening;
  }

  dispose(): void {
    this.stopSpeaking();
    this.recognition?.abort();
  }
}

File: apps/mobile_flutter/lib/core/voice/voice_service.dart
import 'package:flutter_tts/flutter_tts.dart';
import 'package:speech_to_text/speech_recognition_result.dart';
import 'package:speech_to_text/speech_to_text.dart' as stt;

/// Outcome of asking [VoiceService] to start listening.
enum VoiceListenOutcome {
  listening,
  unavailable,
  permissionDenied,
}

/// Outcome of asking [VoiceService] to speak a phrase.
enum SpeakOutcome {
  /// Spoke using a voice that matches the requested language, so
  /// pronunciation should be correct.
  spokenInLanguage,

  /// Spoke, but no voice for the requested language was found, so playback
  /// used the engine's default voice. Pronunciation may not respect the
  /// target language (e.g. an English voice reading Kannada text).
  spokenWithFallbackVoice,

  /// No text-to-speech voice was available at all.
  noVoice,
}

/// Thin wrapper around the device/browser text-to-speech and speech-to-text
/// engines, used to make the lesson screen's play/record buttons actually do
/// something today.
///
/// This is a stopgap built on platform voice APIs (flutter_tts +
/// speech_to_text) rather than a new backend provider. It intentionally does
/// not try to guarantee a consistent voice per learner or route audio
/// through the provider gateway — that is Stage 2/3 work in
/// docs/product/roadmap-learning-features.md. This class exists so "Play
/// phrase" and "Record answer" are functional now instead of no-ops.
class VoiceService {
  VoiceService({FlutterTts? tts, stt.SpeechToText? speech})
      : _tts = tts ?? FlutterTts(),
        _speech = speech ?? stt.SpeechToText();

  final FlutterTts _tts;
  final stt.SpeechToText _speech;
  bool _speechInitialized = false;

  /// Speaks [text] using the requested locale when available, falling back
  /// to the engine's default voice otherwise. Returns a [SpeakOutcome] so
  /// callers can tell a genuine "spoke it correctly" from "spoke it with
  /// whatever voice was available" (e.g. an English voice reading Kannada
  /// script/romanization, which mispronounces it) instead of silently
  /// treating both as success.
  Future<SpeakOutcome> speak(String text, {required String localeId}) async {
    if (text.trim().isEmpty) {
      return SpeakOutcome.noVoice;
    }

    await _tts.stop();
    final hasMatchingVoice = await _hasVoiceForLocale(localeId);
    try {
      await _tts.setLanguage(localeId);
    } catch (_) {
      // Locale not installed/supported; continue with the default voice
      // rather than blocking playback entirely.
    }
    try {
      await _tts.setSpeechRate(0.45);
      final result = await _tts.speak(text);
      if (result != 1) {
        return SpeakOutcome.noVoice;
      }
      return hasMatchingVoice ? SpeakOutcome.spokenInLanguage : SpeakOutcome.spokenWithFallbackVoice;
    } catch (_) {
      return SpeakOutcome.noVoice;
    }
  }

  Future<void> stopSpeaking() => _tts.stop();

  /// Best-effort check for whether any installed TTS voice matches
  /// [localeId] (exact match) or at least its base language (e.g. "kn" for
  /// "kn-IN"). Not all platforms expose `getVoices`, so this fails open
  /// (returns false, meaning "assume no dedicated voice") rather than
  /// throwing.
  Future<bool> _hasVoiceForLocale(String localeId) async {
    try {
      final voices = await _tts.getVoices as List<dynamic>?;
      if (voices == null) {
        return false;
      }
      final languagePrefix = localeId.split('-').first.toLowerCase();
      return voices.any((voice) {
        final voiceLocale = (voice is Map ? voice['locale'] : null)?.toString().toLowerCase();
        if (voiceLocale == null) {
          return false;
        }
        return voiceLocale == localeId.toLowerCase() || voiceLocale.startsWith('$languagePrefix-');
      });
    } catch (_) {
      return false;
    }
  }

  /// Starts listening for speech in [localeId] when the platform supports
  /// it, otherwise falls back to the device default locale. Recognized text
  /// (interim and final) is streamed through [onResult].
  Future<VoiceListenOutcome> startListening({
    required String localeId,
    required void Function(String recognizedWords, bool isFinal) onResult,
  }) async {
    final available = await _ensureSpeechInitialized();
    if (!available) {
      return VoiceListenOutcome.permissionDenied;
    }

    final supportsRequestedLocale = await _localeSupported(localeId);

    try {
      await _speech.listen(
        localeId: supportsRequestedLocale ? localeId : null,
        onResult: (SpeechRecognitionResult result) {
          onResult(result.recognizedWords, result.finalResult);
        },
      );
      return VoiceListenOutcome.listening;
    } catch (_) {
      return VoiceListenOutcome.unavailable;
    }
  }

  Future<void> stopListening() => _speech.stop();

  bool get isListening => _speech.isListening;

  Future<bool> _ensureSpeechInitialized() async {
    if (_speechInitialized) {
      return true;
    }
    try {
      _speechInitialized = await _speech.initialize();
    } catch (_) {
      _speechInitialized = false;
    }
    return _speechInitialized;
  }

  Future<bool> _localeSupported(String localeId) async {
    try {
      final locales = await _speech.locales();
      return locales.any((locale) => locale.localeId == localeId);
    } catch (_) {
      return false;
    }
  }

  void dispose() {
    _tts.stop();
    _speech.stop();
  }
}
